src/array/cpu/spmm_blocking_libxsmm.h

```diff
@@ -0,0 +1,577 @@
+      cur_csr.indptr = IndPtr + M_start;
+      cur_csr.indices = Indices;
+      cur_csr.data = Edges;
+
+      block_csr_array[m] = cur_csr;
+    }
+  }
+}
+
+
+template <typename IdType, typename DType, typename Op>
+inline libxsmm_meltwfunction_opreduce_vecs_idx SpMMCreateLibxsmmKernel(
+    bool has_idx,
+    IdType N,
+    int32_t _ld,
+    libxsmm_meltw_opreduce_vecs_flags redop_flag,
+    bool is_cmp) {
+
+  libxsmm_meltw_opreduce_vecs_flags opredop_flags;
+  if(std::is_same<Op, op::Add<DType>>::value)
+  {
+    opredop_flags = LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OP_ADD;
+  }
+  else if(std::is_same<Op, op::Sub<DType>>::value)
+  {
+    opredop_flags = LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OP_SUB;
+  }
+  else if(std::is_same<Op, op::Mul<DType>>::value)
+  {
+    opredop_flags = LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OP_MUL;
+  }
+  else if(std::is_same<Op, op::Div<DType>>::value)
+  {
+    opredop_flags = LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OP_DIV;
+  }
+  else if(std::is_same<Op, op::CopyLhs<DType>>::value)
+  {
+    opredop_flags = LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OP_COPY;
+  }
+  else if(std::is_same<Op, op::CopyRhs<DType>>::value)
+  {
+    opredop_flags = LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OP_COPY;
+  }
+  if(std::is_same<Op, op::CopyLhs<DType>>::value)
+  {
+    opredop_flags = (libxsmm_meltw_opreduce_vecs_flags)(opredop_flags | LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OPORDER_VECIDX_VECIN);
+  }
+  else if(std::is_same<Op, op::CopyRhs<DType>>::value)
+  {
+    opredop_flags = (libxsmm_meltw_opreduce_vecs_flags)(opredop_flags | LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OPORDER_VECIN_VECIDX);
+    if(!has_idx)
+    {
+      opredop_flags = (libxsmm_meltw_opreduce_vecs_flags)(opredop_flags | LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_IMPLICIT_INDEXED_VECIDX);
+    }
+  }
+  else
+  {
+    opredop_flags = (libxsmm_meltw_opreduce_vecs_flags)(opredop_flags | LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OPORDER_VECIDX_VECIN);
+    if(has_idx)
+    {
+      opredop_flags = (libxsmm_meltw_opreduce_vecs_flags)(opredop_flags | LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_INDEXED_VEC);
+    }
+    else
+    {
+      opredop_flags = (libxsmm_meltw_opreduce_vecs_flags)(opredop_flags | LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_IMPLICIT_INDEXED_VEC);
+    }
+  }
+  opredop_flags = (libxsmm_meltw_opreduce_vecs_flags)(opredop_flags | redop_flag);
+  if(is_cmp)
+  {
+    if(Op::use_lhs)
+    {
+      opredop_flags = (libxsmm_meltw_opreduce_vecs_flags)(opredop_flags | LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_RECORD_ARGOP_OFF_VEC_0);
+    }
+    if(Op::use_rhs)
+    {
+      opredop_flags = (libxsmm_meltw_opreduce_vecs_flags)(opredop_flags | LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_RECORD_ARGOP_OFF_VEC_1);
+    }
+  }
+  libxsmm_meltwfunction_opreduce_vecs_idx kernel=nullptr;
+  if(std::is_same<DType, uint16_t>::value)
+    kernel = libxsmm_dispatch_meltw_opreduce_vecs_idx(N, &_ld, &_ld, LIBXSMM_DATATYPE_BF16, LIBXSMM_DATATYPE_BF16, (sizeof(IdType) == 8) ? LIBXSMM_DATATYPE_I64 : LIBXSMM_DATATYPE_I32, opredop_flags);
```

Comment:
This line might be too long. Our line limit is 100 characters.

Reply:
Added the comments and split the longer lines into multiple lines.